staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go

```diff
@@ -257,10 +276,76 @@ func (a *mutatingDispatcher) callAttrMutatingHook(ctx context.Context, h *v1beta
 		return false, apierrors.NewInternalError(err)
 	}
 
-	changed := !apiequality.Semantic.DeepEqual(attr.VersionedObject, newVersionedObject)
+	changed = !apiequality.Semantic.DeepEqual(attr.VersionedObject, newVersionedObject)
 	trace.Step("Patch applied")
+	if changed {
+		annotator.addPatchAnnotation(result.Patch, result.PatchType)
+	}
 	attr.Dirty = true
 	attr.VersionedObject = newVersionedObject
 	o.GetObjectDefaulter().Default(attr.VersionedObject)
 	return changed, nil
 }
+
+type webhookAnnotator struct {
+	attr                  *generic.VersionedAttributes
+	patchAnnotationKey    string
+	mutationAnnotationKey string
+	webhook               string
+	configuration         string
+}
+
+func newWebhookAnnotator(attr *generic.VersionedAttributes, round, idx int, webhook, configuration string) *webhookAnnotator {
+	return &webhookAnnotator{
+		attr:                  attr,
+		patchAnnotationKey:    fmt.Sprintf("%sround_%d_index_%d", PatchAuditAnnotationPrefix, round, idx),
+		mutationAnnotationKey: fmt.Sprintf("%sround_%d_index_%d", MutationAuditAnnotationPrefix, round, idx),
+		webhook:               webhook,
+		configuration:         configuration,
+	}
+}
+
+func (w *webhookAnnotator) addMutationAnnotation(mutated bool) {
+	if w.attr == nil || w.attr.Attributes == nil {
+		return
+	}
+	value := mutationAnnotationValue(w.configuration, w.webhook, mutated)
+	if err := w.attr.Attributes.AddAnnotation(w.mutationAnnotationKey, value); err != nil {
+		klog.Warningf("failed to set mutation annotation for mutating webhook key %s to %s: %v", w.mutationAnnotationKey, value, err)
+	}
+}
+
+func (w *webhookAnnotator) addPatchAnnotation(patch []byte, patchType admissionv1.PatchType) {
+	if w.attr == nil || w.attr.Attributes == nil {
+		return
+	}
+	var value string
+	var err error
+	switch patchType {
+	case admissionv1.PatchTypeJSONPatch:
+		value, err = jsonPatchAnnotationValue(w.configuration, w.webhook, patch)
+		if err != nil {
+			klog.Warningf("unexpected error composing mutating webhook JSON patch annotation: %v", err)
+			return
+		}
+	default:
+		klog.Warningf("unsupported patch type for mutating webhook annotation: %v", patchType)
+		return
+	}
+	if err := w.attr.Attributes.AddAnnotationWithLevel(w.patchAnnotationKey, value, auditinternal.LevelRequest); err != nil {
+		// NOTE: we don't log actual patch in kube-apiserver log to avoid potentially
+		// leaking information
+		klog.Warningf("failed to set patch annotation for mutating webhook key %s; confugiration name: %s, webhook name: %s", w.patchAnnotationKey, w.configuration, w.webhook)
+	}
+}
+
+func mutationAnnotationValue(configuration, webhook string, mutated bool) string {
+	return fmt.Sprintf(`{"configuration": "%s", "webhook": "%s", "mutated": "%t"}`, configuration, webhook, mutated)
+}
+
+func jsonPatchAnnotationValue(configuration, webhook string, patch []byte) (string, error) {
+	if _, err := jsonpatch.DecodePatch(patch); err != nil {
```

Comment:
if we get to the point of calling this function, can we not assume the patch is well-formed?

Reply:
also, is there a way to avoid decoding the patch twice?

Reply:
is there a max-allowed length for audit annotations?

Reply:
We know the patch is well-formed already when we call this function. This is a redundant check. Dropping this check for performance concern. 

> is there a max-allowed length for audit annotations?

I didn't see one for the `value` from the [documentation](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/) and the source code:
- [record into admission attribute](https://github.com/kubernetes/kubernetes/blob/d951d19ec4d8ec829e83d146200ea91b664b6fa6/staging/src/k8s.io/apiserver/pkg/admission/attributes.go#L128-L144)
- [record into audit event]( https://github.com/kubernetes/kubernetes/blob/d951d19ec4d8ec829e83d146200ea91b664b6fa6/staging/src/k8s.io/apiserver/pkg/audit/request.go#L218-L231)